Нагдалян Арсен. Homework 1

ArsenHD opened the pull request.

ArsenHD committed:
Solution

ArsenHD committed:
Update architecture

ArsenHD committed:
Add some tests

ArsenHD committed:
Move variable substitution to a separate class

ArsenHD committed:
Add more tests

ArsenHD committed:
Add CI script

ArsenHD committed:
Update README.md

ArsenHD committed:
Rework API in order to support infinite input

ArsenHD committed:
Update tests according to the new API

ArsenHD committed:
Remove explicit input specification for commands, because now it is read from an input stream + unite arguments and options

ArsenHD committed:
Support grep command + improve options parsing

ArsenHD committed:
Fix pipeline command connection

ArsenHD committed:
Add tests for grep command

ArsenHD committed:
Add new branch trigger to CI

ArsenHD committed:
Update README.md

ArsenHD committed:
Do not print extra newlines in cat command

FirstTimeInForever commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,28 @@
+on:
+  # Triggers the workflow on push or pull request events but only for the dev-branch branch
+  push:
+    branches: [ dev-branch ]
```
Почему только на `dev-branch`? А `master` или другие бранчи? `-0.2`.

ArsenHD replied:
Теперь на все бранчи с помощью паттерна '**'

FirstTimeInForever commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,28 @@
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+
+      - name: Run Tests
```
Было бы неплохо еще какой-нибудь линтер тут запустить, например `ktlint` и `detekt`. `-0.5`.

ArsenHD replied:
Добавил ktlint

FirstTimeInForever commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,28 @@
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+
+      - name: Run Tests
```
Можно еще покрытие посчитать для развлечения.

ArsenHD replied:
Добавил подсчет с помощью JaCoCo в CI и вставил code coverage badge в readme

FirstTimeInForever commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,28 @@
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+
+      - name: Run Tests
```
А еще было бы неплохо публиковать артефакты.

ArsenHD replied:
Сделал публикацию тест репортов (срабатывает при проваленном пайплайне), а также публикацию fat jar, инструкция по запуску в readme

FirstTimeInForever commented on build.gradle:
```diff
@@ -0,0 +1,35 @@
+plugins {
```
А почему на Groovy, а не Kotlin Script? :(

ArsenHD replied:
Особой причины не было, выбрал дефолт, т.к. проект небольшой и в build.gradle особо много писать не придется

FirstTimeInForever commented on build.gradle:
```diff
@@ -0,0 +1,35 @@
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.7.0'
+}
+
+sourceSets {
```
А чем стандартный лейаут не устроил? (`src/main/kotlin`, `src/test/kotlin`)

ArsenHD replied:
Так получается меньше вложенности. А еще файлы, отвечающие за связанные вещи можно складывать близко друг к другу вне зависимости от языка, в отличие от стандартного лейаута

FirstTimeInForever commented on build.gradle:
```diff
@@ -0,0 +1,35 @@
+
+dependencies {
+    implementation "org.jetbrains.kotlin:kotlin-stdlib"
+    implementation "commons-io:commons-io:2.11.0"
```
А где вы эту зависимость используете?

ArsenHD replied:
Наверное забыл убрать, сейчас удалил

FirstTimeInForever commented on README.md:
```diff
@@ -1 +1,14 @@
-#Software Design HW#1
+# Software Design HW#1
```
Было бы кратко описать что это за проект и высокоуровневый обзор его архитектуры.

ArsenHD replied:
Сделал более подробный readme

FirstTimeInForever commented on README.md:
```diff
@@ -1 +1,14 @@
-#Software Design HW#1
+# Software Design HW#1
```
Так же было бы здорово описать то, что и каких версий нужно Вашему приложению для работы (JDK для сборки и JRE для запуска, например). И самое главное, __как запускать Ваше приложение__ (что писать в терминал?).

ArsenHD replied:
Приложение запускается из fat jar, все зависимости внутри, инструкция по запуску в readme, требования по версии java также указаны

FirstTimeInForever commented on README.md:
```diff
@@ -1 +1,14 @@
+
+### Поддержанные механизмы:
+* Команды
+  * cat
```
Тут надо описать синтаксис этих команд и что они делают.

FirstTimeInForever commented on README.md:
```diff
@@ -1 +1,14 @@
+  * wc
+  * external command (вызов внешнего процесса)
+  * exit (останавливает shell)
+* Пайплайны из команд
```
Нужно добавить синтаксис пайплайнов.

ArsenHD replied:
Сделал более подробное описание синтаксиса всех команд

FirstTimeInForever commented on src/ru/itmo/sd/shell/exception/Exceptions.kt:
```diff
@@ -0,0 +1,14 @@
+
+import ru.itmo.sd.shell.parser.Token
+
+class UndefinedVariableException(name: String) : Exception("Variable '$name' is not defined")
```
Нигде не используется вместе с `IllegalCommandArgumentException`.

ArsenHD replied:
Убрал неиспользуемые исключения

FirstTimeInForever commented on src/ru/itmo/sd/shell/exception/Exceptions.kt:
```diff
@@ -0,0 +1,14 @@
+
+class UnexpectedTokenException(token: Token) : Exception("Token $token was not expected")
+
+class IllegalCommandArgumentException(override val message: String?) : Exception()
```
Почему `Throwable`, а не более узкий тип?

ArsenHD replied:
Конкретно это исключение теперь удалено. Но теперь везде использую Exception (или собственного наследника), потому что эти исключения касаются логики программы (например, некорректный ввод), а не ошибки программиста или JVM

FirstTimeInForever commented on src/ru/itmo/sd/shell/exception/Exceptions.kt:
```diff
@@ -0,0 +1,14 @@
+
+class UnexpectedTokenException(token: Token) : Exception("Token $token was not expected")
+
+class IllegalCommandArgumentException(override val message: String?) : Exception()
```
Почему тут наследуется от `Exception`, а в `ExecutionFailureError` от `Throwable`?

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/util/CliUtils.kt:
```diff
@@ -0,0 +1,35 @@
+        ExecutionResult(code, output.toString(), exception)
+}
+
+internal fun execution(block: ExecutionResultBuilder.() -> Unit): ExecutionResult {
```
Модно, но можно было бы и руками конструировать `ExecutionResult`.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/util/CliUtils.kt:
```diff
@@ -0,0 +1,35 @@
+fun runCommand(command: CliCommand, input: String? = null): String {
+    val (code, output, exception) = command.execute(input)
+    if (code != ReturnCode.OK && exception != null) {
+        println(exception.message)
```
Почему захардкожен `stdout`? Да еще и с ньюлайном в конце.

ArsenHD replied:
Эта функция сейчас уже удалена, но я теперь использую writer кастомного output стрима, чтобы не быть привязанным к stdout

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/Token.kt:
```diff
@@ -0,0 +1,15 @@
+    LET,    // let
+    DASH,   // -
+    ASSIGN, // =
+    END     // <end of input>
```
Было бы логично тут еще иметь токены для кавычек и пробелов.

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/Token.kt:
```diff
@@ -0,0 +1,15 @@
+package ru.itmo.sd.shell.parser
+
+enum class Token {
+    CAT,    // cat
```
Не понятно как команды связаны с типами токенов. Как клиенту Вашего кода добавить новую команду, не добавляя сюда новый тип токена?

ArsenHD replied:
Удалил команды из токенов. Теперь команды создаются с помощью абстрактной фабрики. Нужная фабрика получается по имени команды, а затем с помощью нее создается инстанс команды, в который передаются аргументы, стримы и т.д.

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandLexer.kt:
```diff
@@ -0,0 +1,80 @@
+        private const val WC = "wc"
+        private const val PWD = "pwd"
+        private const val EXIT = "exit"
+        private const val PIPE = "|"
```
Почему бы эти штуки не перенести в енам `Token` как проперти?

ArsenHD replied:
Done

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandParser.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun parseAssignment(): CliVariableAssignment {
+        require(currentToken == Token.LET) { unexpectedToken("'let'") }
```
Что будет если тут вылетит `IAE`? У Вас в вызывающем коде он нигде не поймается.

ArsenHD replied:
Теперь основной цикл защищен обработкой исключений и при ошибке продолжит чтение следующей команды

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandParser.kt:
```diff
@@ -0,0 +1,101 @@
+    private fun parseAssignment(): CliVariableAssignment {
+        require(currentToken == Token.LET) { unexpectedToken("'let'") }
+
+        lexer.nextToken()
```
Вообще, было бы логично написать вспомогательную функцию, которая комбинировала бы `require` и `lexer.nextToken`.

ArsenHD replied:
Сделал. Теперь используется метод Lexer.advance(), опционально принимающий ожидаемый токен и выполняющий проверку

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandParser.kt:
```diff
@@ -0,0 +1,101 @@
+            when (currentToken) {
+                Token.PIPE -> {
+                    lexer.nextToken()
+                    command = PipelineCommand(command, parseSimpleCommand())
```
Почему бы просто не сложить пайплайн в список?

ArsenHD replied:
Теперь складываю в список. Для обеспечения связи между отдельными командами в пайплайне я подставляю в команды PipedInputStream/PipedOutputStream, формирующие цепочку и запускаю каждую команду в отдельном потоке. Пайплайн завершается, когда все команды заканчивают работу.

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandParser.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun parseSimpleCommand(): CliSimpleCommand {
+        val buildCommand = when (currentToken) {
```
Так, это совсем не годится. Смотрите, сейчас чтобы добавить новую команду Вам нужно:
1. Добавить новый элемент в `Token`
2. Добавить его обработку в `CommandLexer`
3. Добавить конструктор этой команды в это свич

Явно можно сделать это проще:
1. Вам нужно избавиться от захардкоженых команд (убрать токены для них из `Token` и научить лексер и парсер жить без них).
2. Вам нужно посмотреть на строку и понять что это команда и у нее такие-то аргументы (в лексере или тут в парсере).
3. Логично что парсер не должен знать про то как нужно выполнять команды.
4. Вам нужна фабрика. А еще лучше абстрактная фабрика. Эта фабрика умеет по строковому названию команды конструировать имплементации ваших команд.
5. Этой фабрикой нужно параметризировать какую-то сущность которая умеет выполнять команды (например, Ваш `CommandProcessor`), но никак не парсер.

Более того, с таким подходом клиенты вашего кода смогут добавлять новые команды (думаю понятно почему в текущем варианте это невозможно).
`-2`.

ArsenHD replied:
В другом комментарии уже описывал исправление. Убрал все лишнее из енама Token и сделал абстрактную фабрику, как Вы посоветовали

FirstTimeInForever commented on src/ru/itmo/sd/shell/parser/CommandParser.kt:
```diff
@@ -0,0 +1,101 @@
+import ru.itmo.sd.shell.exception.UnexpectedTokenException
+
+class CommandParser {
+    private lateinit var lexer: CommandLexer
```
Хорошая практика параметризировать парсер лексером (нужно сделать `interface Lexer`, `CommandLexer: Lexer` и сделать `CommandParser` зависимым от `Lexer`).

ArsenHD replied:
Создал интерфейс лексера и храню его объект в парсере.

FirstTimeInForever commented on src/ru/itmo/sd/shell/processor/CommandProcessor.kt:
```diff
@@ -0,0 +1,44 @@
+import ru.itmo.sd.shell.parser.CommandParser
+
+class CommandProcessor {
+    private val environment: Environment = Environment()
```
Тут можно не указывать типы.

ArsenHD replied:
Done

FirstTimeInForever commented on src/ru/itmo/sd/shell/processor/CommandProcessor.kt:
```diff
@@ -0,0 +1,44 @@
+import ru.itmo.sd.shell.parser.CommandParser
+
+class CommandProcessor {
+    private val environment: Environment = Environment()
```
Почему бы не параметризировать `CommandProcessor` всеми этими штуками?

FirstTimeInForever commented on src/ru/itmo/sd/shell/processor/CommandProcessor.kt:
```diff
@@ -0,0 +1,44 @@
+
+    fun run() {
+        var line = readLine()
+        while (line != null) {
```
В этом методе должна быть какая-нибудь обработка ошибок.

ArsenHD replied:
Теперь ловлю все исключения, обрабатываю некоторым образом и продолжаю чтение следующей команды

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/CliElements.kt:
```diff
@@ -0,0 +1,45 @@
+import java.io.ByteArrayInputStream
+import java.io.PipedInputStream
+
+sealed interface CliElement
```
Как клиентам расширять `sealed`?

ArsenHD replied:
Убрал модификатор `sealed`

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/CliElements.kt:
```diff
@@ -0,0 +1,45 @@
+ */
+sealed class CliSimpleCommand : CliCommand() {
+    abstract val name: String
+    abstract val options: List<Option>
```
У вас в куче мест `options` и `arguments` переопределяются с `emptyList()`. Почему бы не сделать такую дефолтную имплементацию?

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/CliElements.kt:
```diff
@@ -0,0 +1,45 @@
+        return processStdin()
+    }
+
+    open fun processArguments(): ExecutionResult = throw UnsupportedOperationException()
```
Брутально. Почему бы не конструировать пустой `ExecutionResult`?

ArsenHD replied:
Этот код был сильно переработан. Теперь логика исполнения команды реализуется одним методом execute(). Команды кастомизируются стримами ввода вывода, поэтому им не важно откуда читать и куда писать, все работает через общий интерфейс обработки стримов.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExitCommand.kt:
```diff
@@ -0,0 +1,18 @@
+
+    override fun execute(input: String?): ExecutionResult {
+        println("*** Exiting shell ***")
+        exitProcess(0)
```
Ух, это совсем жестко. С таким подходом вызывающий код не получит назад управление. Не получится, например, освободить какие-нибудь захваченные ресурсы. `-0.4`.

ArsenHD replied:
Теперь сигнализирую об окончании работы с помощью ShellShutdownException. Он ловится и shell заканчивает работу, предварительно освободив все ресурсы. Исключения в целом не стоит использовать для control flow, но остановка shell происходит ровно один раз, поэтому перформанс не должен проседать.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/PwdCommand.kt:
```diff
@@ -0,0 +1,17 @@
+    override val name: String = "pwd"
+
+    override fun execute(input: String?): ExecutionResult = execution {
+        output.appendLine(System.getProperty("user.dir"))
```
С учетом того, что для `cd`/`ls` нужно будет изменять значение этого проперти, это плохая практика - прочитайте документацию к `System.getProperty`. Было бы логично поддерживать своё собственное значение `cwd`, у Вас как раз есть удачное место для этого - `Environment`. `-0.2`.

ArsenHD replied:
Добавил свойство `currentDirectory` в `Environment` и использую здесь его

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/PipelineCommand.kt:
```diff
@@ -0,0 +1,11 @@
+
+class PipelineCommand(val left: CliCommand, val right: CliSimpleCommand) : CliCommand() {
+    override fun execute(input: String?): ExecutionResult {
+        val leftOutput = runCommand(left, input)
```
Если хранить пайплайн списком, не придется думать о рекурсии.

ArsenHD replied:
Теперь храню списком. Уже описывал как это работает в одном из комеентариев

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+    override val name: String = command
+
+    override fun execute(input: String?): ExecutionResult = execution {
+        val process = Runtime.getRuntime().exec("${this@ExternalCommand}")
```
Почему не `ProcessBuilder`?

ArsenHD replied:
Теперь использую ProcessBuilder

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+    override fun execute(input: String?): ExecutionResult = execution {
+        val process = Runtime.getRuntime().exec("${this@ExternalCommand}")
+        val returnCode = process.waitFor()
+        val error = process.errorStream.reader().readText()
```
Стримы нужно использовать с `try-with-reources`. В котлине для этого есть `Closable.use`.

ArsenHD replied:
Теперь использую use.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+        val process = Runtime.getRuntime().exec("${this@ExternalCommand}")
+        val returnCode = process.waitFor()
+        val error = process.errorStream.reader().readText()
+        if (error.isNotEmpty()) {
```
Так у вас потеряется обычный output команды.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+        val process = Runtime.getRuntime().exec("${this@ExternalCommand}")
+        val returnCode = process.waitFor()
+        val error = process.errorStream.reader().readText()
+        if (error.isNotEmpty()) {
```
Нужно проверять не `stderr`, а `exit code`. `stderr` по дефолту вообще не выводится в шелл. `-0.5`.

ArsenHD replied:
Теперь проверяю exit code. Если он `!= 0`, то кидаю исключение, которое ловится в основном цикле.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+            code = returnCode
+            return@execution
+        }
+        output.append(process.inputStream.reader().readText())
```
Команда могла завершиться с пустым `stderr`, но ненулевым `exit code`. Кажется, `exit code` в таком случае потеряется.

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/ExternalCommand.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    override fun toString(): String {
+        val opts = options.joinToString(" ") { "-${it.value}" }
```
На самом деле `options` и `arguments` это две одинаковые вещи. Много команд/приложений ожидают их в том порядке, в котором они были написаны в терминале. Тут у Вас, насколько я понимаю, этот порядок может нарушаться. `-0.2`.

ArsenHD replied:
Теперь и опции и аргументы хранятся в `arguments`, и команда знает какие у нее есть опции, поэтому может извлечь их из списка аргументов по имени

FirstTimeInForever commented on src/ru/itmo/sd/shell/cli/command/CatCommand.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private fun processFiles() = execution {
+        val files = arguments.map { File(it) }
```
Лучше использовать более современный `nio.Path`.

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/AbstractCommandTest.kt:
```diff
@@ -0,0 +1,18 @@
+        val testFiles = Utils.testFiles
+        val testDirs = Utils.testDirs
+
+        val filesContent: Map<String, String> =
```
Думаю, это лишнее - в тестах достаточно каждый раз читать файл, вместо того чтобы занимать память их содержимым (у Вас могут быть очень большие файлы).

ArsenHD replied:
Убрал

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/CatCommandTest.kt:
```diff
@@ -0,0 +1,38 @@
+        @JvmStatic
+        fun argumentProvider(): List<Arguments> {
+            return listOf(
+                arguments(testFiles.subList(0, 3)),
```
Это интересные `subList`... Обычно более наглядным и более поддерживаемым решением является явно прописать нужные вам файлы.

ArsenHD replied:
Сделал тестовые данные более явными, в том числе файлы

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/CatCommandTest.kt:
```diff
@@ -0,0 +1,38 @@
+        @JvmStatic
+        fun argumentProvider(): List<Arguments> {
+            return listOf(
+                arguments(testFiles.subList(0, 3)),
```
Плюс, что делать если у Вас добавятся новые тестовые данные или поменяется порядок файлов?

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/CatCommandTest.kt:
```diff
@@ -0,0 +1,38 @@
+        val cat = command(fileNames)
+        val expected = expectedCat(fileNames)
+        val (code, output) = cat.execute()
+        assertEquals(0, code)
```
У Вас в нескольких тестах этот метод почти одинаковый. Может быть стоит его как-нибудь обобщить?

ArsenHD replied:
Обобщил запуск команды и проверку результата с помощью `AbstractCommandTest.assertResultSuccessful()`

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/ExternalCommandTest.kt:
```diff
@@ -0,0 +1,38 @@
+class ExternalCommandTest : AbstractSimpleCommandTest() {
+
+    @Test
+    fun testExternalCommand() {
```
Тест неплохой, но `ls` может не существовать на каких-то платформах. Стоит подумать как пофиксить.

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/PipelineCommandTest.kt:
```diff
@@ -0,0 +1,28 @@
+import org.junit.jupiter.api.Test
+import ru.itmo.sd.shell.cli.util.ExecutionResult
+
+class PipelineCommandTest : AbstractCommandTest() {
```
Надо бы тут еще пару тестов написать с большим количеством команд в пайплайне.

ArsenHD replied:
Добавил тестов

FirstTimeInForever commented on test/ru/itmo/sd/shell/cli/command/WcCommandTest.kt:
```diff
@@ -0,0 +1,51 @@
+    @ParameterizedTest
+    @MethodSource("argumentProvider")
+    fun testWc(input: List<String>, expected: String) {
+        mockkStatic(CONSOLE_KT)
```
Этого не пришлось бы делать, если бы ваш шелл умел параметризироваться io стримами и использовал их для команд и всего остального io.

ArsenHD replied:
Теперь CommandProcessor параметризуется стримами. Он парсит команды и передает в них эти стримы.

FirstTimeInForever reviewed (changes requested):
По функционалу:
* Не работает:
```
let x = ex
let y = it
$x$y
```
`-0.3`

* `cat` не умеет читать файлы с расширениями:
```
cat gradlew.bat
.bat
cat: .bat: No such file or directory
```
`-0.3`

* В `echo some | cat` выводится лишний ньюлайн `-0.1`

* `wc` не умеет читать файлы с расширениями:
```
wc gradlew.bat
.bat
wc: .bat: No such file or directory
```
`-0.3`

* `wc` неправильно считает количество байт:
```
# yours
wc gradlew
185 783 5589 gradlew
# coreutils
# notice 5774 = 5589 + 185 :)
wc gradlew
     185     783    5774 gradlew
```
`-0.2`

* Пайп в `wc` явно работает неправильно (сравните с башем):
```
echo 123 | wc
2 1 3
```

* С пайпами или `cat` явно что-то не так:
```
echo 123 | cat | cat | cat | wc
5 1 3
echo 123 | cat | cat | cat | cat | wc
6 1 3
```
`-0.5`

* Двойные и одинарные кавычки работают неправильно:
```
let x = 1
echo $x
1
echo "$x"
" $x "
echo '$x'
' $x '
```
`-1`

* Неправильно обрабатывается пустая переменная:
```
echo $a
$a
```
`-0.25`

* Никак не обрабатываются несбалансированные кавычки (или не делается так как в баше):
```
echo "asd
"asd
echo 'asd
'asd
```
`-0.25`

* Вызов `cat` без аргументов не кидает ошибку (или не ведёт себя как в баше)
`-0.25`

* Всё падает при пайпе в пустоту:
```
cat asd |
Exception in thread "main" ru.itmo.sd.shell.exception.UnexpectedTokenException: Token END was not expected
	at ru.itmo.sd.shell.parser.CommandParser.parseSimpleCommand(CommandParser.kt:72)
	at ru.itmo.sd.shell.parser.CommandParser.parseCliCommand(CommandParser.kt:54)
	at ru.itmo.sd.shell.parser.CommandParser.parse(CommandParser.kt:28)
	at ru.itmo.sd.shell.processor.CommandProcessor.run(CommandProcessor.kt:18)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt:7)
	at ru.itmo.sd.shell.ShellKt.main(Shell.kt)
```
`-0.25`

Остальное:
* Надо поправить `README.md`. `-1`.
* У Вас нет `.gitignore` - стоит добавить.
* Было бы неплохо добавить адекватный `.editorcofig`.
* Стоит научить Ваш шелл работать не только с `stdin`/`stdout`, но и с произвольными стримами (параметризируйте ваш шелл `InputStream` и `OutputStream`). Это Вам сильно поможет в тестировании.
* Реализовав прошлый пункт можно (и стоит) будет написать интеграционный тест - просимулировать ввод нескольких команд и проверить их вывод.

Кажется, у Вас очень много достаточно несложно исправляется, так что не переживайте.
`0.4/10`.

ArsenHD committed:
Fix extra newlines in echo command

ArsenHD committed:
Fix variable substitution + brackets processing

ArsenHD committed:
Make commands use bytewise read/write operations

ArsenHD committed:
Write error messages to a separate error stream

ArsenHD committed:
Simplify execution result API

ArsenHD committed:
Fix: remove double quotes from input in all cases

ArsenHD committed:
Use abstract factory to create command objects

This makes parser independent from the set of supported commands

ArsenHD committed:
Modify lexer and parser to support multiline input

ArsenHD committed:
Remove unnecessary property from ExternalCommand

ArsenHD committed:
Make ExecutionFailureError an inheritor of Exception

ArsenHD committed:
Parametrize parser by lexer

ArsenHD committed:
Move command factory up in package hierarchy

ArsenHD committed:
Make CliElement not sealed to allow external users to extend commands hierarhy

ArsenHD committed:
Allow application to finish gracefully upon exit command

ArsenHD committed:
Make empty list a default value for command arguments

ArsenHD committed:
Parametrize commands with IO streams + store piped commands in lists

ArsenHD committed:
Minor formatting fix

ArsenHD committed:
Print errors to output like bash does

ArsenHD committed:
Fix creation of process for external command

ArsenHD committed:
Stop 'wc' command execution when end of transmission character is encountered

ArsenHD committed:
Add IO streams customization to command factories

ArsenHD committed:
Allow all commands to have customized IO streams

ArsenHD committed:
Use executor service to run pipeline commands in threads

ArsenHD committed:
Use executor service to run pipeline commands in threads

ArsenHD committed:
Refactor parser instantiation

ArsenHD committed:
Parse commands with customized IO streams

ArsenHD committed:
Fix main loop error isolation

ArsenHD committed:
Refactor tests: use factories for command instantiation, better arguments, move out common code

ArsenHD committed:
Add parser tests

ArsenHD committed:
Allow input reading to be interrupted when the pipeline is complete, but the input stream is not

ArsenHD committed:
Do not close the given IO streams, but rather the created reader and writer

The IO streams should be handled by the one who created them, not by the command

ArsenHD committed:
Remove unused exceptions

ArsenHD committed:
Add environment to command and let all types of cli elements be executed

ArsenHD committed:
Make test data more clear

ArsenHD committed:
Forbid unmatched quotes

ArsenHD committed:
Add 'private' modifier to a property

ArsenHD committed:
Add .gitignore

ArsenHD committed:
Add .editorconfig for Kotlin code

ArsenHD committed:
Add linter to CI + trigger CI on all branches

ArsenHD committed:
Minor code clean up

ArsenHD committed:
Save test run results in artifacts

ArsenHD committed:
Finish reading remaining bytes after interruption

ArsenHD committed:
Add code coverage computation to CI

ArsenHD committed:
Add more tests on parsing and a test on wc command with files

ArsenHD committed:
Use JaCoCo to compute code coverage

ArsenHD committed:
Make wc test work on linux

ArsenHD committed:
Fix JaCoCo .csv file location

ArsenHD committed:
Fix code coverage badge generation

ArsenHD committed:
Add coverage badge to README.md

ArsenHD committed:
Add fatJar task and upload CI artifacts

ArsenHD committed:
Make README.md more detailed

ArsenHD committed:
Use unix-style line separator in gradlew and run.sh

ArsenHD committed:
Use custom output stream for messages to the user instead of always using stdout

ArsenHD committed:
Print a start message when the interpreted is started

ArsenHD committed:
Hide gradle build output when running shell with a script run.sh

ArsenHD committed:
Add current directory info to environment

ArsenHD committed:
Make program free all its resources when it exits by any reason

ArsenHD committed:
Remove unused dependency

ArsenHD reviewed (commented):
Исправил код в соответствии с замечаниями. Также починил неработающую функциональность (кавычки, многострочные пайплайны, некорректную работу пайплайнов, подстановок переменных и т.д.). Также добавил .gitignore, .editorconfig, code coverage badge, дополнил README.

Отмечу, что некоторая часть исправлений уже присутствовала в решении ДЗ-2, поэтому я сделал cherry pick из ветки ДЗ-2 и добавил новые исправления.

При запуске через `./run.sh` нужно дождаться сообщения `*** The interpreter has been started ***`. До этого скрипт выполняет билд jar файла. Конечно, в первый раз это займет какое-то время, а затем будет работать быстрее.

ArsenHD committed:
Autogenerated JaCoCo coverage badge

FirstTimeInForever commented:
Сломался вызов внешних команд (`-1`).
Смотрите, вы не совсем правильно поняли идею паттерна абстракная фабрика. Основная идея паттерна - спрятать логику создания сущностей. Для нас тут релевантна возможность менять набор команд, которые может поддерживать шелл. У Вас сейчас можно легко подменять имплементации отдельных команд, но сам набор команд все равно жестко забит. В Вашем случае этот же паттерн нужно было тогда так же применить для `CommandFactoryHandler`. В идеале должно быть вот так:

```kotlin
interface Command

interface CommandFactory {
  fun createCommand(...): Command
}

open class DefaultCommandFactory: CommandFactory {
  override fun createCommand(...): Command {
    return when (...) {
      "cat" -> CatCommand(...)
      ... -> ...
      else -> error(...)
    }
  }
}

// Consumers:

class CommandParser(private val commandFactory: CommandFactory) {
	...

	fun parse() {
		val command = commandFactory.createCommand(...)
		...
	}
}

// Single point for creating all app components
fun main() {
	// To extend/change commands just swap factory implementation
	val factory = DefaultCommandFactory()
	val parser = CommandParser(factory, ...)
	...
}
```
`-0.5`
Остально считаю исправленным. Получилось норм. 
`8.5/10`.
